weave/trace_server/trace_server_common.py

```diff
@@ -1,6 +1,90 @@
 import copy
+import datetime
+import json
 from collections import OrderedDict
-from typing import Any, Dict, Optional
+from typing import Any, Dict, Literal, Optional, Union, cast
+
+from weave.trace_server import trace_server_interface as tsi
+from weave.trace_server.refs_internal import InternalObjectRef, parse_internal_uri
+
+
+def make_derived_summary_fields(
+    call_dict: Dict[str, Any], summary_key: Literal["summary", "summary_dump"]
+) -> tsi.SummaryMap:
+    """
+    Make derived summary fields for a call.
+
+    Summary is controlled by the user, but the `weave` summary key is
+    used to store derived fields, adhering to the tsi.SummaryMap type.
+
+    Args:
+        call_dict: The call dict.
+        summary_key: The key in the call dict that contains the summary. In
+            the clickhouse server this is "summary_dump", but in sqlite
+            it's "summary".
+
+    Returns:
+        The derived summary fields.
+    """
+    display_name = call_dict.get("display_name")
+    started_at_dt = _make_datetime_from_any(call_dict["started_at"])
+    ended_at_dt = _make_datetime_from_any(call_dict.get("ended_at"))
+
+    status = "success"
+    if call_dict.get("exception"):
+        status = "error"
+    elif ended_at_dt is None:
+        status = "running"
+
+    latency = None
+    if ended_at_dt and started_at_dt:
+        latency = (ended_at_dt - started_at_dt).microseconds
+
+    if not display_name:
+        op = parse_internal_uri(call_dict["op_name"])
+        if isinstance(op, InternalObjectRef):
+            display_name = op.name
+        else:
+            display_name = call_dict["op_name"]
+
+    summary = _load_json_maybe(call_dict.get(summary_key)) or {}
+
+    weave_summary = summary.get("weave", {})
+    weave_summary["trace_name"] = display_name
+    weave_summary["status"] = status
+    if latency is not None:
+        weave_summary["latency"] = latency
+
+    summary["weave"] = weave_summary
+    return cast(tsi.SummaryMap, summary)
```

Comment:
instead of casting, can you directly construct this?

Reply:
I can't seem to construct this while keeping mypy happy, because while we control the `weave` key in the summary, there are also arbitrary user controlled keys. `tsi.SummaryMap` allows this, but mypy doesn't know that...